docs/src/appendix/mast.md

```diff
@@ -0,0 +1,528 @@
+
+```rust
+pub trait MerkleTreeNode {
+    fn digest(&self) -> Digest;
+}
+
+pub trait Decorated {
+    fn decorators(&self) -> &DecoratorList;
+    fn decorate(&mut self, decorators: DecoratorList);
+}
+```
+
+Let's start with the _basic block_, which is what comprises the majority of leaf nodes in the MAST:
+
+```rust
+#[derive(MerkleTreeNode, Decoratable)]
+pub struct BasicBlockNode {
+    /// The primitive operations contained in this basic block.
+    ///
+    /// The operations are broken up into batches of 8 groups,
+    /// with each group containing up to 9 operations, or a
+    /// single immediates. Thus the maximum size of a basic block
+    /// is 72 operations. A larger block would need to be encoded
+    /// in MAST as a join of two basic blocks.
+    batches: Vec<OpBatch<8>>,
+}
+
+pub struct OpBatch<const BATCH_SIZE: usize = 8> {
+    ops: Vec<Operation>,
+    groups: [Felt; BATCH_SIZE],
+    op_counts: [usize; BATCH_SIZE],
+    num_groups: usize,
+}
+```
+
+As you can see, this node type has no other nodes as children, instead it has a finite set of `Operation`, which are the primitive operations corresponding to Miden's instruction set. Basically what you'd expect in a basic block.
+
+Next, let's take a look at the `Join` node type, which is used for scheduling. What do I mean by scheduling? Consider the following Rust program:
+
+```
+pub fn start(x: usize, y: usize) {
+    let a = x * y;
+    let b = x + y;
+    let c = a + b;
+    println!("{c}");
+}
+```
+
+While the code implies that `let a = x * y` will be executed before `let b = x + y;`, that isn't necessarily the case. In fact, the compiler can emit code to execute these expressions in any order it likes. We refer to a specific ordering, a _schedule_, and the work of constructing a schedule, _scheduling_.
+
+With that out of the way, let's see how `Join` is used to impose a schedule on the control flow graph of a program. First, the control flow graph:
+
+```mermaid
+graph TD
+    A["x * y"]
+    B["x + y"]
+    C["a + b"]
+    D["println"]
+    A->C
+    B->C
+    C->D
+```
+
+Which could be represented in MAST as the following (assuming for a moment that basic blocks don't exist, and each primop is its own node type):
+
+```mermaid
+graph TD
+    A["JOIN"]
+    B["mul"]
+    C["JOIN"]
+    D["add"]
+    E["add"]
+    F["JOIN"]
+    G["println"]
+    A->B
+    A->C
+    C->D
+    C->F
+    F->E
+    F->G
+```
+
+So, with all of that in mind, let's look at the definition of `JoinNode`:
+
+```rust
+#[derive(MerkleTreeNode)]
+pub struct JoinNode {
+    children: [MastNodeId; 2],
+}
+```
+
+Not much to it!
+
+The `Split` node type is functionally identical in representation, but unlike `Join`, is not a scheduling node, but a control flow node. It only executes one of its children, depending on a boolean value on top of the operand stack.
+
+```rust
+#[derive(MerkleTreeNode)]
+pub struct SplitNode {
+    children: [MastNodeId; 2],
+}
```

Comment:
I noticed that none of the node structs have `hash` fields - is the assumption that they would be computed dynamically by implementing `MerkleTreeNode` trait? If so, we may need to rethink this strategy as hash computations will likely be by far the most time consuming part. So, ideally, these hashes would be serialized as well so that we don't have to re-compute them on de-serialization.

Reply:
To clarify, I wasn't trying to imply that we actually remove the `hash` field from the structures/serialized form, but factor out the repetitive details of each node type as part of the `MerkleTreeNode` trait. In other words, how the trait is implemented is an unimportant detail, what is important is that every node type _has_ a hash.

As you've pointed out, the perf wins make memoization of the hash as part of the node structure essential.